### PR TITLE
Caption feature

### PR DESCRIPTION
**Pull request by Comandeer**
## What is the purpose of this pull request?

New feature for new feature

## Does your PR contain necessary tests?

All patches which change the editor code must include tests. You can always read more
on [PR testing](https://docs.ckeditor.com/ckeditor4/docs/#!/guide/dev_contributing_code-section-tests),
[how to set the testing environment](https://docs.ckeditor.com/ckeditor4/docs/#!/guide/dev_tests) and
[how to create tests](https://docs.ckeditor.com/ckeditor4/docs/#!/guide/dev_tests-section-creating-your-own-test)
in the official CKEditor documentation.

### This PR contains

- [x] Unit tests
- [x] Manual tests

## What changes did you make?

I've added caption feature, which is responsible for toggling visibility of empty caption and for showing placeholder text.

### CODE REVIEW

File: plugins/imagebase/plugin.js [hunk 1]
```diff
@@ -44,7 +66,84 @@
...
+			 * @member CKEDITOR.plugins.imagebase.featuresDefinitions.caption
+			 * @param {CKEDITOR.dom.element} sender Element, which triggered `selectionChange` event
+			 */
+			_toggleCaption: function( sender ) {
+				var isFocused = getFocusedWidget( this.editor ) === this,
+					caption = this.parts.caption,
+					editable = this.editables.caption;
+
+				if ( isFocused ) {
+					caption.removeAttribute( 'data-cke-hidden' );
```
Comment (mlewand) [inline plugins/imagebase/plugin.js]: This attribute should be removed _after_ the placeholder was changed. This would mean less reflows which is always a good thing 👍

**Comment by Comandeer** [comment]
I'm still investigating failing tests in IE and Edge. Something fishy's going there as manual test for the same case is working just fine (probably artificial selection change is not working).

I also don't know how to deal with `getData` issue. I was trying to extend filter in `editor.dataProcessor`, but it seems that fetched data does not have all needed attributes (no `data-cke-*` ones), so I can't really filter out placeholder. Asserting its content is risky as user could have caption "Enter image caption". The only idea I have is to move placeholder content into CSS generated content.

**Comment by Comandeer** [comment]
I've pushed `t/932-2992-css` with working approach with placeholder inserted via CSS generated content. It fixed also issue with failing test in `uploadwidget` suite.

Tests in IE and Edge are still failing. it seems that these browsers don't fire `_refreshCaption` after inserting content into caption programatically. I didn't find a viable solution yet.

**Comment by Comandeer** [comment]
I've pushed improvement for tests both on `t/932-2992` and `t/932-2992-css`. Now tests pass on Edge. One test is still failing on IE: ` test toggling captions (two widgets with empty captions) (classic)` – but only if it's run with other tests and when console is closed.

**Comment by Comandeer** [comment]
I've pushed fix to both of branches.

**Comment by mlewand** [comment]
I have a feeling that this is an edge case, but I'll put it here for my future reference, so that it can be extracted to a separate issue.

## Provide detailed reproduction steps (if any)

1. Open [Easy Image sample](http://ckeditor.test/plugins/easyimage/samples/easyimage.html).
1. Make sure it's an updated version, where CKEditor old logo is displayed under the header.
1. Move your mouse over the widget and out of it.

### Expected result

Widget wrapper gets highlighted with yellow outline.

### Actual result

Widget wrapper is highlighted, but weird reflows are taking place.

## Other details

* Browser: Chrome
* OS: Windows 10

HIDPI scaling: 250%

I wasn't able to repro it on other browsers, but I think to very specific layouting in this case (the cause for it might be that widget drag handler is affecting DOM - scroll bar appears, the width changes and reflow occurs.

File: tests/plugins/imagebase/features/caption.js [hunk 2]
```diff
@@ -0,0 +1,334 @@
...
+				assertVisibility( caption, options.blur, 'caption visibility (blur)' );
+
+				if ( options.onBlur ) {
+					options.onBlur( widget );
+				}
+			} );
+		};
+	}
+
+	function createMultipleToggleTest( options ) {
```
Comment (mlewand) [inline tests/plugins/imagebase/features/caption.js]: Please, add at least minimal comment to what the function does/what is its purpose. Currently one has to read its source code to understand this.

File: tests/plugins/imagebase/features/caption.js [hunk 3]
```diff
@@ -0,0 +1,334 @@
...
+						} );
+					}, 50 );
+
+					widget.focus();
+					wait();
+				} );
+			}
+
+			bot.setData( getFixture( options.fixture ), function() {
+				var widgets = getWidgets( editor, options.widgetsCount );
```
Comment (mlewand) [inline tests/plugins/imagebase/features/caption.js]: Wouldn't it be a better idea to use widely available API? just go with `editor.widgets.instances.slice()` instead of adding new redundant code.

File: tests/plugins/imagebase/features/caption.js [hunk 4]
```diff
@@ -0,0 +1,334 @@
...
+
+			function assertMultipleVisibility( widgets, expected, msg ) {
+				forEach( widgets, function( widget, i ) {
+					assertVisibility( widget.parts.caption, expected[ i ],
+						'caption#' + i + ' visibility (' + msg + ')' );
+				} );
+			}
+
+			function assertOnFocus( widgets, expected ) {
+				forEach( widgets, function( widget, i ) {
```
Comment (mlewand) [inline tests/plugins/imagebase/features/caption.js]: In reality this code will never execute assertions for more than 1 widget. In YUI `wait` throws an exception, so the flow is  never resumed afterwards.

File: tests/plugins/imagebase/features/caption.js (hunk 2 above)
Comment (mlewand) [reply tests/plugins/imagebase/features/caption.js]: Please, review this assertion. It's a cause for `'test toggling captions (two widgets with empty captions)'` to fail on IE11.

File: plugins/easyimage/plugin.js [hunk 5]
```diff
@@ -198,6 +198,8 @@
 			widgetDefinition = CKEDITOR.plugins.imagebase.addFeature( editor, 'link', widgetDefinition );
```
[the remark this answers is not in the document]
Comment (mlewand) [reply plugins/easyimage/plugin.js]: After I rebased it onto latest `t/932-2961` branch - that brought Balloon Panel, and it has an issue. I'll add a manual test for that shortly.

File: tests/plugins/imagebase/features/manual/caption.html [hunk 6]
```diff
@@ -0,0 +1,34 @@
+<div id="classic">
```
Comment (mlewand) [inline tests/plugins/imagebase/features/manual/caption.html]: Please, add inline editors. You  have added `floatingspace` plugin in `.md` file but you did not use it with any editor.

File: plugins/easyimage/plugin.js (hunk 5 above)
[the remark this answers is not in the document]
Comment (mlewand) [reply plugins/easyimage/plugin.js]: There's other issue related to blurring:

1. Focus a EI widget with no caption, so that placeholder text appears.
2. Put the focus outside of the editor.

## Expected

Caption is removed/hidden.

## Actual

Caption + placeholder remains visible.

Added manual test for that in [tests\plugins\imagebase\features\manual\captionblur.md](https://github.com/ckeditor/ckeditor-dev/blob/8115f8d/tests/plugins/imagebase/features/manual/captionblur.md#L5) test.

File: tests/plugins/imagebase/features/caption.js (hunk 2 above)
Comment (Comandeer) [reply tests/plugins/imagebase/features/caption.js]: @mlewand which assertion exactly are you thinking of?

File: plugins/imagebase/plugin.js [hunk 7]
```diff
@@ -44,7 +70,103 @@
...
+			widget.parts.caption.data( 'cke-placeholder', true );
+			widget.editables.caption.setData( widget.editor.lang.imagebase.captionPlaceholder );
+		}
+
+		function removePlaceholder( widget ) {
+			widget.parts.caption.data( 'cke-placeholder', false );
+			widget.editables.caption.setData( '' );
+		}
+
+		function toggleVisibility( caption, isVisible ) {
```
[the remark this answers is not in the document]
Comment (Comandeer) [reply plugins/imagebase/plugin.js]: It seems that you don't like word "toggle". Every time you demand to change it to something else :D

File: plugins/imagebase/plugin.js [hunk 8]
```diff
@@ -44,7 +70,103 @@
...
+			caption.data( 'cke-active', isVisible );
```
[the remark this answers is not in the document]
Comment (Comandeer) [reply plugins/imagebase/plugin.js]: I've also renamed `data-cke-placeholder` to `data-cke-caption-placeholder` for consistency.

File: tests/plugins/imagebase/features/caption.js (hunk 2 above)
Comment (mlewand) [reply tests/plugins/imagebase/features/caption.js]: Actually I was referring to `createMultipleToggleTest` helper.

File: plugins/imagebase/plugin.js (hunk 7 above)
[the remark this answers is not in the document]
Comment (mlewand) [reply plugins/imagebase/plugin.js]: Hehe, I like toggle methods, but only when it actually _toggles_ something 😉

**Comment by mlewand** [comment]
@Mgsy can I ask you for help with testing this feature?

**Comment by Mgsy** [comment]
## Steps to reproduce
1. Select the image without the caption.
2. Apply `Bold`.
3. Focus the image.
4. Focus the caption.

## Current result
Placeholder doesn't disappear.

## Expected result
Placeholder disappears.

## Other information
OS: Windows 10, MacOS X
Browser: All browsers

**Comment by Mgsy** [comment]
## Steps to reproduce
1. Focus the image.
2. Copy it.
3. Put the caret inside image caption.
4. Paste.

## Current result
Image has been pasted, but editor has crashed.

## Expected result
Image is pasted without errors.

## Error
```
plugin.js:2858 Uncaught TypeError: Cannot read property 'editables' of undefined
    at plugin.js:2858
    at CKEDITOR.htmlParser.element.forEach (fragment.js:622)
    at CKEDITOR.htmlParser.element.forEach (fragment.js:636)
    at CKEDITOR.htmlParser.element.forEach (fragment.js:636)
    at CKEDITOR.htmlParser.element.forEach (fragment.js:636)
    at Editor.<anonymous> (plugin.js:2813)
    at Editor.listenerFirer (event.js:144)
    at Editor.CKEDITOR.event.fire (event.js:290)
    at Editor.CKEDITOR.editor.fire (editor_basic.js:24)
    at CKEDITOR.htmlDataProcessor.toDataFormat (htmldataprocessor.js:294)
```

## Other information
OS: Windows 10, MacOS X
Browser: All browsers

**Comment by Mgsy** [comment]
## Steps to reproduce
1. Focus the empty image caption.
2. Apply `Bold`.

## Current result
Caption disappears and selection moves to the previous block.

## Expected result
Selection stays inside the caption and `bold` activates.

## Other information
OS: Windows 10, MacOS X
Browser: Firefox, Edge

**Comment by Mgsy** [comment]
## Steps to reproduce
1. Focus the empty caption.
2. Type `test`.
3. Undo.

## Current result
Caption disappears.

## Expected result
Caption is visible and selection stays inside the empty caption.

## Other information
OS: Windows 10, MacOS X
Browser: All browsers

**Comment by Mgsy** [comment]
## Steps to reproduce
1. Open the sample with 2 images with captions.
2. Focus first image's caption.
3. Press <kbd>Tab</kbd>.
4. Select caption of second image with <kbd>Ctrl</kbd> + <kbd>A</kbd>.
5. Press <kbd>Delete</kbd>.
6. Press <kbd>Shift</kbd> + <kbd>Tab</kbd>.

## Current result
Second image's empty caption is visible. 

## Expected result
Second image's empty caption hides.

## Other information
OS: Windows 10, MacOS X
Browser: All browsers

**Comment by Comandeer** [comment]
I've extracted issues raised by @Mgsy as separate subtasks. Additionally I've reported #1469.

File: plugins/imagebase/plugin.js [hunk 9]
```diff
@@ -186,7 +216,108 @@
...
+
+				if ( !this.parts.caption ) {
+					this.parts.caption = createCaption( this );
+				}
+
+				this._refreshCaption();
+			},
+
+			/**
+			 * Method used inside {@link CKEDITOR.editor#event-selectionChange} event to decide if caption for given widget
```
Comment (mlewand) [inline plugins/imagebase/plugin.js]: You have tied this method strongly with the `selectionChange` and now it uses also `blur`. It's possible there will be different events soon. Please make the docs event-agnostic.

File: tests/plugins/easyimage/manual/balloontoolbar/caption.md [hunk 10]
```diff
@@ -0,0 +1,26 @@
...
+
+	Balloon panel shows, again in the same position.
+
+	### Unexpected
+
+	Balloon panel points to the bottom edge **of an image**, overlapping the caption.
+
+---
+
+Repeat above steps for all other editors.
```
Comment (mlewand) [inline tests/plugins/easyimage/manual/balloontoolbar/caption.md]: Please, mention that Edge users could be affected by #1458 - and experience focus quirks when focusing between the editors.

**Comment by mlewand** [comment]
1. Open http://tests.ckeditor.test:1030/tests/plugins/imagebase/features/manual/caption
1. Focus the first captionless EI widget.
1. Cut it (ctrl/cmd+x).
1. Paste it.

### Expected result

Widget is focused, caption placeholder is visible.

### Actual result

Widget is focused but there's no caption placeholder visible.

## Other details

* Browser: Chrome, Edge, Firefox
* OS: Windows 10

File: tests/plugins/imagebase/features/manual/caption.html [hunk 11]
```diff
@@ -0,0 +1,92 @@
+<h2>Classic editor</h2>
```
Comment (mlewand) [inline tests/plugins/imagebase/features/manual/caption.html]: Now that we have inline editors, please import CSS files, so that the widgets in inline editors are styled properly.

File: tests/plugins/imagebase/features/caption.js [hunk 12]
```diff
@@ -0,0 +1,455 @@
+/* bender-tags: editor,widget */
```
Comment (mlewand) [inline tests/plugins/imagebase/features/caption.js]: Please ignore this test in unspported environments.

File: plugins/imagebase/plugin.js [hunk 13]
```diff
@@ -186,7 +216,108 @@
...
+		}
+
+		return {
+			setUp: function( editor ) {
+				var listeners = [];
+
+				function listener( evt ) {
+					var path = evt.name === 'blur' ? editor.elementPath() : evt.data.path,
+						sender = path ? path.lastElement : null,
+						widgets = editor.widgets,
```
[the remark this answers is not in the document]
Comment (Comandeer) [reply plugins/imagebase/plugin.js]: I don't see any 115 chars line, there is only 135 chars one, but it's not the previous one 🤔

[the remark this answers is not in the document]
Comment (mlewand) [reply plugins/imagebase/plugin.js]: I'm sorry, wasn't accurate. What I meant is that `widgets` variable is used only in line `previous = widgets.getByElement( editor.editable(…)` then the rest is applicable.

File: tests/plugins/imagebase/features/manual/caption.html (hunk 11 above)
Comment (Comandeer) [reply tests/plugins/imagebase/features/manual/caption.html]: Actually there aren't any styles that could be additionally loaded – all are already loaded. It looks differently than EI because it's a custom, test image widget:

**Comment by Comandeer** [comment]
I've fixed all raised issues, I've also rebased to the newest `t/932-2961`.

File: tests/plugins/imagebase/features/manual/caption.html (hunk 11 above)
Comment (mlewand) [reply tests/plugins/imagebase/features/manual/caption.html]: I mean, it's a figure - I wanted figures to be styled consistently in all editors. It just requires importing `contents.css` - see 289bd705b.
